### ISSUE TITLE
Allow multiple hosting startup assemblies to be specified without requiring semi-colon shenanigans

### ISSUE THREAD

**Issue by pranavkm** [dotnet/aspnetcore#19303]
This came up as part of https://github.com/dotnet/aspnetcore/issues/17272. Rather than relying on semi-colons to separate multiple listing, it would be convenient if multiple hosting startup assemblies could be keyed off a common prefix rather than requiring the `ASPNETCORE_HOSTINGSTARTUPASSEMBLIES` key. Consider:

```
ASPNETCORE_HOSTINGSTARTUPASSEMBLIES = "AssemblyA;AssemblyB"

vs

ASPNETCORE_HOSTINGSTARTUPASSEMBLIES_KeyA = "AssemblyA"
ASPNETCORE_HOSTINGSTARTUPASSEMBLIES_KeyB = "AssemblyB"
```

In particular when authoring templates, this avoids the accidentally stomping over an existing value or writing complicated if-elses.

**Comment by anurse** [comment]
@pranavkm what's the problem with `;`? All the other similar lists (in the .NET Core host, etc.) use `;`-separated lists. It's a pretty standard practice so I'm not sure why we would deviate from it.

**Comment by anurse** [comment]
Also, maybe we should have a chat about how runtime compilation is plugging in. In general, we'd prefer to use Startup Hooks (the feature built-in to the .NET host) instead of hosting startup.

**Comment by pranavkm** [comment]
@anurse the problem was that if you need to enable two hosting startup assemblies, you have to be careful to write code that appends rather than overwriting content. 

Authoring conditions in our template is problematic since you'd have to riddle it with ifdefs. Similarly if values came from two different sources, such as file + environment, you're it's difficult to combine these values.

**Comment by anurse** [comment]
> the problem was that if you need to enable two hosting startup assemblies, you have to be careful to write code that appends rather than overwriting content.

This is true for a lot of variables that affect us, not just `HOSTINGSTARTUPASSEMBLIES`. Startup hooks work this way, `PATH`, etc. Also, the value is ordered, which can matter, so I'm not sure how we'd do that in the prefix model you propose.

**Comment by anurse** [comment]
We talked about this last week but I don't remember *exactly* what we landed on. I think we were suggesting that it would be better to have the places that specify values for these support "recursive" expansion, for example:

```json
{
	"ASPNETCORE_HOSTINGSTARTUPASSEMBLIES": "A.dll;%ASPNETCORE_HOSTINGSTARTUPASSEMBLIES%;B.dll"
}
```

cc @DamianEdwards @pranavkm

**Comment by pranavkm** [comment]
Do we have room to squeeze this in 5.0? If we were, I'd like to make sure we update the MVC templates to account for this

**Comment by anurse** [comment]
> Do we have room to squeeze this in 5.0?

Well, the suggested change in my previous comment isn't something we'd do, it would be in VS (if it doesn't actually work now, which it actually might...).

Making a change to how hosting reads this variable is not in scope for 5.0 at this time, we don't have resources.
